### ISSUE TITLE
file system watcher takes too much CPU.

### ISSUE THREAD

**Issue by stevej** [BuoyantIO/linkerd#391]
Looking at the profile (/admin/pprof/profile), we see `LinuxWatchService.poll` being our number 1 cpu user. That is unexpected.

```
Using local file ./profile_before_slowness_16.
Welcome to pprof!  For help, type 'help'.
(pprof) top
Total: 2125 samples
     394  18.5%  18.5%      394  18.5% sun.nio.fs.LinuxWatchService.poll
     287  13.5%  32.0%      303  14.3% org.jboss.netty.buffer.DynamicChannelBuffer.ensureWritableBytes
     142   6.7%  38.7%      142   6.7% java.lang.Thread.currentThread
```

Once we hit our throughput degradation issue, it drops to number 2 but it is still an issue.

```
stevej@proxy-test-4:~/linkerd$ google-pprof ./profile_after_slowness_16
Using local file ./profile_after_slowness_16.
Using local file ./profile_after_slowness_16.
Welcome to pprof!  For help, type 'help'.
(pprof) top
Total: 4694 samples
     755  16.1%  16.1%      755  16.1% java.lang.Thread.currentThread
     746  15.9%  32.0%      746  15.9% sun.nio.fs.LinuxWatchService.poll
     372   7.9%  39.9%      392   8.4% org.jboss.netty.buffer.DynamicChannelBuffer.ensureWritableBytes
     257   5.5%  45.4%      257   5.5% sun.misc.Unsafe.compareAndSwapObject
     210   4.5%  49.9%      223   4.8% java.util.Arrays.copyOfRange
```

With oliver's help, I added some logging to `namer/fs/src/main/scala/io/buoyant/namer/fs/Watcher.scala` and didn't see anything unusual.

```
I 0509 21:48:32.785 THREAD1: /metrics => io.buoyant.linkerd.admin.MetricsHandler$
I 0509 21:48:32.787 THREAD1: Serving admin http on 9990
I 0509 21:48:32.922 THREAD1: serving int on /0.0.0.0:4140
E 0509 21:49:27.192 THREAD24: fs observing disco
E 0509 21:49:27.333 THREAD37: fs waiting for events on disco
```

Our best guess is that the way we're calling the API is not quite right or there's something not tuned about a default linux VM (this is using debian-8 on GCE) that causes us to spin too much. We really need to reduce this number or use a different technique in the fsWatcher. 18% of the CPU at high load is about half a core and we're expecting people to use this in small containers.

**Comment by stevej** [comment]
Based on a [blog post](http://blog.omega-prime.co.uk/?p=161), this is the Design of the watcher, there's a Thread running that busy-waits on `poll`

[Source of poll from jdk7 (this seems unchanged in 8)](https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/solaris/classes/sun/nio/fs/LinuxWatchService.java#L314)

And evidence that we're sitting waiting for poll.
```
"35" : {
      "priority" : 5,
      "state" : "RUNNABLE",
      "daemon" : true,
      "thread" : "Thread-3",
      "stack" : [
        "sun.nio.fs.LinuxWatchService.poll(Native Method)",
        "sun.nio.fs.LinuxWatchService.access$600(LinuxWatchService.java:47)",
        "sun.nio.fs.LinuxWatchService$Poller.run(LinuxWatchService.java:314)",
        "java.lang.Thread.run(Thread.java:745)"
      ]
```

Are we just hosed using the filesystem watcher? Should we push people harder to use Service Discovery? We don't intend for the fs watcher to be used full time. Maybe we should just document this behavior?

**Comment by stevej** [comment]
We could use a different java [inotify library](https://bitbucket.org/nbargnesi/inotify-java) but would we have to release a different binary for linux?

**Comment by stevej** [comment]
How to reproduce: startup l5d and send a single request to kick the FS watcher into action.

**Comment by stevej** [comment]
I threw together a quick branch that used `poll(timeout)` instead of `take()` and it made no difference, we're still busywaiting in a native method.

**Comment by olix0r** [comment]
WONTFIX.  Doc update tracked in #661
